usr/src/uts/armv8/os/fakebop.c

```diff
@@ -332,6 +415,328 @@ do_bsys_alloc(bootops_t *bop, caddr_t virthint, size_t size, int align)
 	return (virthint);
 }
 
+static void
+bsetprop(int flags, char *name, int nlen, void *value, int vlen)
+{
+	uint_t size;
+	uint_t need_size;
+	bootprop_t *b;
+
+	/*
+	 * align the size to 16 byte boundary
+	 */
+	size = sizeof (bootprop_t) + nlen + 1 + vlen;
+	size = (size + 0xf) & ~0xf;
+	if (size > curr_space) {
+		need_size = (size + (MMU_PAGEOFFSET)) & MMU_PAGEMASK;
+		curr_page = do_bsys_alloc(NULL, 0, need_size, MMU_PAGESIZE);
+		curr_space = need_size;
+	}
+
+	/*
+	 * use a bootprop_t at curr_page and link into list
+	 */
+	b = (bootprop_t *)curr_page;
+	curr_page += sizeof (bootprop_t);
+	curr_space -=  sizeof (bootprop_t);
+	b->bp_next = bprops;
+	bprops = b;
+
+	/*
+	 * follow by name and ending zero byte
+	 */
+	b->bp_name = curr_page;
+	bcopy(name, curr_page, nlen);
+	curr_page += nlen;
+	*curr_page++ = 0;
+	curr_space -= nlen + 1;
+
+	/*
+	 * set the property type
+	 */
+	b->bp_flags = flags & DDI_PROP_TYPE_MASK;
+
+	/*
+	 * copy in value, but no ending zero byte
+	 */
+	b->bp_value = curr_page;
+	b->bp_vlen = vlen;
+	if (vlen > 0) {
+		bcopy(value, curr_page, vlen);
+		curr_page += vlen;
+		curr_space -= vlen;
+	}
+
+	/*
+	 * align new values of curr_page, curr_space
+	 */
+	while (curr_space & 0xf) {
+		++curr_page;
+		--curr_space;
+	}
+}
+
+static void
+bsetprops(char *name, char *value)
+{
+	bsetprop(DDI_PROP_TYPE_STRING, name, strlen(name),
+	    value, strlen(value) + 1);
+}
+
+static void
+bsetprop32(char *name, uint32_t value)
+{
+	bsetprop(DDI_PROP_TYPE_INT, name, strlen(name),
+	    (void *)&value, sizeof (value));
+}
+
+static void
+bsetprop64(char *name, uint64_t value)
+{
+	bsetprop(DDI_PROP_TYPE_INT64, name, strlen(name),
+	    (void *)&value, sizeof (value));
+}
+
+static void
+bsetpropsi(char *name, int value)
+{
+	char prop_val[32];
+
+	(void) snprintf(prop_val, sizeof (prop_val), "%d", value);
+	bsetprops(name, prop_val);
+}
+
+/*
+ * to find the size of the buffer to allocate
+ */
+int
+do_bsys_getproptype(bootops_t *bop __unused, const char *name)
+{
+	bootprop_t *b;
+
+	for (b = bprops; b != NULL; b = b->bp_next) {
+		if (strcmp(name, b->bp_name) != 0)
+			continue;
+		return (b->bp_flags);
+	}
+
+	return (-1);
+}
+
+/*
+ * to find the size of the buffer to allocate
+ */
+int
```

Comment:
should this be `size_t`?

Reply:
I can update it - it's transposed from i86pc at the moment.

Reply:
Going to leave this as-is for now. i86pc has it as int as well, and almost all of the callsites expect a signed return value (for error checking).

Perhaps `ssize_t`?